pkg/utils/resources/resources.go

```diff
@@ -125,6 +125,18 @@ func MaxResources(resources ...v1.ResourceList) v1.ResourceList {
 	return resourceList
 }
 
+// ZeroResources returns a default resource list with zeroed resources
+func ZeroResources() v1.ResourceList {
```

Comment:
This is definitely better, but still misses possible resources that could exist (e.g. gpus). I wonder if we could analyze/cache the results of GetInstanceTypes() to hydrate the full set of values available in the cloud provider.

Reply:
Is `GetInstanceTypes` filtered by requirements set in the nodepool? If so, could be a good way to track real values of resources that a nodepool _might_ provision.

Reply:
It's not filtered by requirements. The reason we pass NodePool through is that the NodeClassReference and KubeletConfig in the NodePool _might_ impact available memory/cpu. 

As for allowing the controller to make a call to `GetInstanceTypes()`, I think it's fine after reviewing the discussion. The cloudprovider generally caches this so it shouldn't have a tangible impact on TPS, but we should probably cache the resources that we get here from a call for something like 12h and then re-poll when that cache expires

Reply:
Why do we need another cache?

Reply:
Seems like you can't know what the Cloud Provider is doing when you call GetInstanceTypes() so it's probably best practice to reduce your calls to that interface if you can.

Reply:
I'll take a look this week at updating the PR to use info from `GetInstanceTypes()` so we can support a wider variety of resources

Reply:
Curious Jonathan, why isn't the caching already in the vendor neutral impl?

Reply:
> why isn't the caching already in the vendor neutral impl

I think the real answer is that there hasn't been a proven need for it yet but practically you have to push a lot of the staling mechanisms that we have in the AWS cloudprovider side into the neutral caching mechanism. It's definitely something to explore, I don't disagree. There's definitely also a path to explore here around reconciling the instance types into state and referring to that rather than doing an imperative-based cache. cc: @engedaam